extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java

```diff
@@ -176,6 +180,47 @@ protected KafkaSupervisorSpec toggleSuspend(boolean suspend)
     );
   }
 
+  /**
+   * Extends {@link SeekableStreamSupervisorSpec#validateSpecUpdateTo} to ensure that the proposed spec and current spec are either both multi-topic or both single-topic.
+   * <p>
+   * getSource() returns the same string (exampleTopic) for "topicPattern=exampleTopic" and "topic=exampleTopic".
+   * This override prevents this case from being considered a valid update.
+   * </p>
+   * @param proposedSpec the proposed supervisor spec
+   * @throws DruidException if the proposed spec is not a Kafka spec or if the proposed spec changes from multi-topic to single-topic or vice versa
+   */
+  @Override
+  public void validateSpecUpdateTo(SupervisorSpec proposedSpec) throws DruidException
+  {
+    if (!(proposedSpec instanceof KafkaSupervisorSpec)) {
+      throw InvalidInput.exception(
+          StringUtils.format("Cannot evolve to [%s] from [%s]", getClass().getName(), proposedSpec.getClass().getName())
+      );
+    }
+    KafkaSupervisorSpec other = (KafkaSupervisorSpec) proposedSpec;
+    if (this.getSpec().getIOConfig().isMultiTopic() != other.getSpec().getIOConfig().isMultiTopic()) {
+      throw InvalidInput.exception(getIllegalInputSourceUpdateErrorMessage("(%s) %s", "(%s) %s"),
+                                   this.getSpec().getIOConfig().isMultiTopic() ? "multi-topic" : "single-topic",
+                                   this.getSource(),
+                                   other.getSpec().getIOConfig().isMultiTopic() ? "multi-topic" : "single-topic",
+                                   other.getSource());
+    }
+
+    super.validateSpecUpdateTo(proposedSpec);
+  }
+
+  @Override
+  protected String getIllegalInputSourceUpdateErrorMessage(String existingSource, String proposedSource)
```

Comment:
We shouldn't need to override this method.
We can either use the constant defined in `SeekableStreamSupervisorSpec` as is or a different String altogether.
A cleaner way to include the type name in that String would be to have a placeholder for type and fill it with `getType()` while formatting.